parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java

```diff
@@ -904,4 +932,208 @@ public byte[] getDictPageAAD() {
       return this.dictPageAAD;
     }
   }
+
+  private void processBlocksFromReaderWithStitching(IndexCache indexCache) throws IOException {
+    // TODO add the test for empty files joins, it should merge schemas
+    LOG.info("Rewriting input fileLeft: {}, remaining filesLeft: {}", reader.getFile(), inputFiles.size());
+    int rowGroupIdx = 0;
+    List<BlockMetaData> blocks = reader.getFooter().getBlocks();
+    for (BlockMetaData blockMetaData: blocks) {
+      writer.startBlock(blockMetaData.getRowCount());
+
+      // Writing the left side
+      indexCache.setBlockMetadata(blockMetaData);
+      List<ColumnChunkMetaData> chunksL = blockMetaData.getColumns();
+      for (ColumnChunkMetaData chunk : chunksL) {
+        if (chunk.isEncrypted()) { // TODO add that detail to docs
+          throw new IOException("Column " + chunk.getPath().toDotString() + " is encrypted");
+        }
+        ColumnDescriptor descriptorL = descriptorsMap.get(chunk.getPath());
+        if (descriptorL != null) { // descriptorL might be NULL if a column is from the right side of a join
+          reader.setStreamPosition(chunk.getStartingPos());
+          BloomFilter bloomFilter = indexCache.getBloomFilter(chunk);
+          ColumnIndex columnIndex = indexCache.getColumnIndex(chunk);
+          OffsetIndex offsetIndex = indexCache.getOffsetIndex(chunk);
+          writer.appendColumnChunk(descriptorL, reader.getStream(), chunk, bloomFilter, columnIndex, offsetIndex);
+        }
+      }
+
+      // Writing the right side
+      for (RightColumnWriter writer: columnWritersR) {
```

Comment:
Is it possible to move this into `processBlocksFromReader`?

Reply:
Yeah, that would make sense, let me try.

Reply:
done